From 630121f0029fe554edafa011f09e7f581dddb888 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Wed, 23 Feb 2022 05:57:38 +0200
Subject: [PATCH] [SW-73084] greco: mask i2c2 comb irq in early init

On enabling I2C2 slave during preboot, interrupt is generated
to ARC0 which is trying to initialize PCIE link during cold
reset. This creates exception in ARC0 flow as it is running
bare metal code which cannot handle irqs. To avoid exception
in ARC0 this change masks I2C 2 comb irq (0x47).

Change-Id: Idccad076ee3155e108a29a8a3c1421a60c08e2bf
(cherry picked from commit 4b24edaa5f5fa56b4e2d0c9c1defad95c7683d88)
---

diff --git a/arch/greco/arc_elf32/common.c b/arch/greco/arc_elf32/common.c
index 41b14d8..f243644 100644
--- a/arch/greco/arc_elf32/common.c
+++ b/arch/greco/arc_elf32/common.c
@@ -138,6 +138,14 @@
 	regmap_writel(ARC_RST_VEC_ADDR, mmPSOC_ARC0_CFG_RST_VEC_ADDR);
 	regmap_writel(ARC_RST_VALUE, mmPSOC_ARC0_CFG_ARC_RST);
 
+	/*
+	 * On enabling I2C2 slave during preboot, interrupt is generated
+	 * to ARC0 which is trying to initialize PCIE link during cold
+	 * reset. To avoid exception in ARC0 this change masks I2C 2
+	 * comb irq (0x47).
+	 */
+	regmap_writel(ARC_I2C_2_IRQ_MASK, mmPSOC_ARC0_CFG_IRQ_INTR_MASK_2);
+
 	/* Clear DCCM memory */
 	init_arc_mem(0, ARC0_DCCM_FULL_BASE, ARC0_DCCM_SIZE);
 }
diff --git a/arch/greco/arc_elf32/common.h b/arch/greco/arc_elf32/common.h
index 672f43c..a1fc5d0 100644
--- a/arch/greco/arc_elf32/common.h
+++ b/arch/greco/arc_elf32/common.h
@@ -52,6 +52,7 @@
 
 #define ARC_RST_VALUE		0x110
 #define ARC_RST_VEC_ADDR	0x0
+#define ARC_I2C_2_IRQ_MASK	0x80
 
 #ifndef __ASSEMBLY__
 extern void secondary_entry(void);
